### DISCUSSION TITLE
Standard method to communicate with multiple slaves?

### DISCUSSION THREAD

**Discussion by atlesg** [eModbus/eModbus#187] [Q&A]
Hi guys, I am currently trying communicate with up to 8 slaves IDs (mostly sensors). My response usually ignore some of the slave ID, I don't know why. So I tried many ways: reduce modbus RTU reading interval, read in sequence but nothing works at all. Please give me some advices on this:

My pseudo code:

```
void init_modbus(void)
{
   
    Serial2.begin(9600, SERIAL_8N1, RXPIN, TXPIN);
    RS485.onDataHandler(&handleData);
    RS485.onErrorHandler(&handleError);
    RS485.setTimeout(2000); // Default 2000ms
    RS485.begin();
}
void read_modbus(void)
{
    data_ready = false;
    RS485.addRequest(Sensor_1.modbus_ht_addr, Sensor_1.modbus_ht_addr, READ_HOLD_REGISTER, 0x01, 4);
    RS485.addRequest(Sensor_2.modbus_ht_addr, Sensor_2.modbus_ht_addr, READ_HOLD_REGISTER, 0x01, 4);
    //....
    RS485.addRequest(Sensor_8.modbus_ht_addr, Sensor_8.modbus_ht_addr, READ_HOLD_REGISTER, 0x01, 4);
}
void handleError(Error error, uint32_t token)
{
    // ModbusError wraps the error code and provides a readable error message for it
    ModbusError me(error);
    Serial.printf("Error from Device: %d\n", token);
    Serial.printf("Error response: %02X - %s\n", error, (const char *)me);
}
void handleData(ModbusMessage msg, uint32_t token)
{
    printf("Response: serverID=%d, FC=%d, Token=%08X, length=%d:\n", msg.getServerID(), msg.getFunctionCode(), token, msg.size());
    for (auto &byte : msg)
    {
    Serial.printf("%02X ", byte);
    }
    Serial.println("");
    switch (msg.getServerID())
    {
    case ADDRESS_HT_SENSOR_1:
        msg.get(3, Sensor_1.ht);
        break;
    case ADDRESS_HT_SENSOR_2:
        msg.get(3, Sensor_2.ht);
        break;
   //...
    case ADDRESS_HT_SENSOR_8:
        msg.get(3, Sensor_8.ht);
        break;
    default:
        break;
    }
    data_ready = true;
}
```

Thank you!

**Comment by bertmelis** [comment]
Do you get an error message?
Because you should at least get a timeout message if there is no response.

**Reply by atlesg** [reply]
Yeah my error message is time_out. Should I reduce the timeout from 2000ms to some number? Does this parameter affect the communication much?

**Comment by Miq1** [comment]
Your code snippet looks okay as much as one can see from it. I assume ``Sensor_1`` and ``ADDRESS_HT_SENSOR_1`` etc. have the same values?

Do you have some logs of the servers ignoring their ids? Are you sure to have properly configured these in the servers?

**Reply by atlesg** [reply]
Yes, they have the same value, I already checked multiple times. However, it's really weird that if I read, for example: sensor 1 2 3 4, then esp32 usually ignore the 2 and 4. If I only read 2 or 4, they are okay. So I think there are some problems with the queing, speed or anything...

**Reply by Miq1** [reply]
Your requests will be processed one by one from the request queue. There is no overlap between one request/response and the next.
What could explain it would be if you have set the queue limit to 2 - did you?

**Reply by atlesg** [reply]
I see, I just tried setting the queue limit and went back. Better not touch it I guess. What about the time_out? do you think that I should decrease it?

**Reply by Miq1** [reply]
No, the timeout is just the time to wait after a request to state a nonresponding server. As I wrote the next request will be held until the completion of the previous.

**Reply by Miq1** [reply]
BTW, the queue limit is safe to modify at will, a larger queue only may take longer to process if full. But attempting to put another request onto an already full queue will return immediately from the ``addRequest`` call with an error telling you so.

**Reply by atlesg** [reply]
I see, now I think I understand more of your library! :D

**Comment by Miq1** [comment]
Glad to read you got the knack!

**Reply by atlesg** [reply]
Problem not solved yet xD! I can now read all the slaves but the speed is too slow, like average 2-3 seconds for a slave at Baudrate 9600 (I can't change this Baudrate because it's default for my sensors)

**Reply by Miq1** [reply]
Please post a log of the requests and responses so that we can see what is going on. The times you wrote are rather looking like some server is not responding, so the client will wait for the set timeout time of 2000ms before proceeding to the next.
